File: packages/health_connector_hc_android/lib/src/mappers/health_record_mappers/energy_burned/basal_metabolic_rate_record_mapper.dart
import 'package:health_connector_core/health_connector_core_internal.dart'
    show BasalMetabolicRateRecord, HealthRecordId, sinceV3_6_0, Power;
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/health_record_id_mapper.dart';
import 'package:health_connector_hc_android/src/mappers/metadata_mappers/metadata_mapper.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart'
    show BasalMetabolicRateRecordDto;
import 'package:meta/meta.dart' show internal;

/// Converts [BasalMetabolicRateRecord] to [BasalMetabolicRateRecordDto].
@sinceV3_6_0
@internal
extension BasalMetabolicRateRecordToDto on BasalMetabolicRateRecord {
  BasalMetabolicRateRecordDto toDto() {
    return BasalMetabolicRateRecordDto(
      id: id.toDto(),
      time: time.millisecondsSinceEpoch,
      zoneOffsetSeconds: zoneOffsetSeconds,
      metadata: metadata.toDto(),
      kilocaloriesPerDay: rate.inKilocaloriesPerDay,
    );
  }
}

/// Converts [BasalMetabolicRateRecordDto] to [BasalMetabolicRateRecord].
@sinceV3_6_0
@internal
extension BasalMetabolicRateRecordDtoToDomain on BasalMetabolicRateRecordDto {
  BasalMetabolicRateRecord toDomain() {
    return BasalMetabolicRateRecord.internal(
      id: id?.toDomain() ?? HealthRecordId.none,
      time: DateTime.fromMillisecondsSinceEpoch(time, isUtc: true),
      zoneOffsetSeconds: zoneOffsetSeconds,
      metadata: metadata.toDomain(),
      rate: Power.kilocaloriesPerDay(kilocaloriesPerDay),
    );
  }
}

File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/health_record_mappers/BasalMetabolicRateRecordMapper.kt
package com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers

import androidx.health.connect.client.records.BasalMetabolicRateRecord
import androidx.health.connect.client.units.Power
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.BasalMetabolicRateRecordDto
import java.time.Instant
import java.time.ZoneOffset

/**
 * Converts a Health Connect [BasalMetabolicRateRecord] object to a [BasalMetabolicRateRecordDto].
 */
internal fun BasalMetabolicRateRecord.toDto(): BasalMetabolicRateRecordDto =
    BasalMetabolicRateRecordDto(
        id = metadata.id,
        time = time.toEpochMilli(),
        zoneOffsetSeconds = zoneOffset?.totalSeconds?.toLong(),
        metadata = metadata.toDto(),
        kilocaloriesPerDay = basalMetabolicRate.inKilocaloriesPerDay,
    )

/**
 * Converts a [BasalMetabolicRateRecordDto] to a Health Connect [BasalMetabolicRateRecord] object.
 */
internal fun BasalMetabolicRateRecordDto.toHealthConnect(): BasalMetabolicRateRecord =
    BasalMetabolicRateRecord(
        basalMetabolicRate = Power.kilocaloriesPerDay(kilocaloriesPerDay),
        time = Instant.ofEpochMilli(time),
        zoneOffset = zoneOffsetSeconds?.let { ZoneOffset.ofTotalSeconds(it.toInt()) },
        metadata = metadata.toHealthConnect(id),
    )
